README.rst

django-weasypdf
===============

A Django class-based view helper to generate PDF with WeasyPrint.

Requires: `WeasyPrint <https://github.com/Kozea/WeasyPrint>`_

Optional requirements:

- matplotlib (to render plots)

.. contents::

.. sectnum::

Installation
------------

Install the package by running:

.. code:: bash

    pip install django-weasypdf

or:

.. code:: bash

    pip install git+https://github.com/morlandi/django-weasypdf


You will probably build your own app, for example `reports`, to provide custom derived Views
and templates:

.. code:: bash

    python manage.py startapp reports

In your settings, add:

.. code:: python

    INSTALLED_APPS = [
        ...
        'reports',
        'weasypdf',
    ]

Note that `reports` is listed before `weasypdf` to make sure you can possibly
override any template.

In your urls, add:

.. code:: python

    from django.urls import path, include

    urlpatterns = [
        ...
        path('reports/', include('reports.urls', namespace='reports')),
        ...

Optionally, you might want to copy the default templates from 'weasypdf/templates/weasypdf' to 'reports/templates/reports'
for any required customization; see `Customizing the templates`_ below

A sample report
---------------

A test view has already been included in the `weasypdf` app
to render a sample document for demonstration purposes.

To use it, add the following to your urls:

.. code:: python

    path('weasypdf/', include('weasypdf.urls', namespace='weasypdf')),

then, visit:

    http://127.0.0.1:8000/weasypdf/test/print/


Usage
-----

You can copy the following to your `reports` app to have a sample view
to start working with:

file `reports/urls.py`:

.. code:: python

    from django.urls import path
    from . import views

    app_name = 'weasypdf'

    urlpatterns = [
        path('test/print/', views.ReportTestView.as_view(), {'for_download': False, 'lines': 200, }, name="test-print"),
        path('test/download/', views.ReportTestView.as_view(), {'for_download': True, 'lines': 200, }, name="test-download"),
    ]


file `reports/views.py`:

.. code:: python

    from weasypdf.views import WeasypdfView


    class ReportView(WeasypdfView):

        #my_custom_data = None
        header_template_name = 'weasypdf/header.html'
        footer_template_name = 'weasypdf/footer.html'
        styles_template_name = 'weasypdf/styles.css'

        def get_context_data(self, **kwargs):
            context = super(ReportView, self).get_context_data(**kwargs)
            #self.my_custom_data = context.pop('my_custom_data', None)
            # context.update({
            #     'footer_line_1': config.REPORT_FOOTER_LINE_1,
            #     'footer_line_2': config.REPORT_FOOTER_LINE_2,
            # })
            return context


    class ReportTestView(ReportView):
        body_template_name = 'weasypdf/pages/test.html'
        styles_template_name = 'weasypdf/pages/test.css'
        # header_template_name = None
        # footer_template_name = None
        title = "Report Test"

        def get_context_data(self, **kwargs):
            context = super().get_context_data(**kwargs)

            # Add a plot
            try:
                from .plot import build_plot_from_data
                plot_image = build_plot_from_data(data=None, as_base64=True)
                context.update({
                    'plot_image': plot_image,
                })
            except:
                pass

            # Add your stuff here ...
            context.update({
                ...
            })

            return context


or **replace `weasypdf/header.html` with `reports/header.html`**, etc ... when using
custom templates.

file `reports/pages/test.html`:

.. code:: html

    {% extends "weasypdf/base.html" %}

    {% block content %}

        <h1>Test PDF</h1>

        {% if plot_image %}
            <img class="plot" src="data:image/png;base64,{{plot_image}}">
        {% endif %}

        {% with lines=lines|default:100 %}
            {% for i in "x"|rjust:lines %}
                <div>line {{forloop.counter}} ...</div>
            {% endfor %}
        {% endwith %}

    {% endblock content %}


You can now download the PDF document at:

    http://127.0.0.1:8000/reports/test/download/

or open it with the browser at:

    http://127.0.0.1:8000/reports/test/print/

You can inspect the HTML used for PDF rendering by appending `?format=html` to the url:

    http://127.0.0.1:8000/reports/test/print/?format=html

.. image:: screenshots/pdf_sample.png


Building a PDF document from a background process
-------------------------------------------------

A `WeasypdfView.render_as_pdf_to_stream(self, base_url, extra_context, output)` method is supplied for this purpose:

.. code:: python

    def render_as_pdf_to_stream(self, base_url, extra_context, output):
        """
        Build the PDF document and save in into "ouput" stream.

        Automatically called when the view is invoked via HTTP (unless self.format == 'html'),
        but you can also call it explicitly from a background task:

            view = PdfTestView()
            context = view.get_context_data()
            with open(filepath, 'wb') as f:
                view.render_as_pdf_to_stream('', context, f)
        """

A sample management command to build a PDF document outside the HTML request/response
cycle is available here:

`weasypdf/management/commands/build_test_pdf.py <./weasypdf/management/commands/build_test_pdf.py>`_


Providing "extra_context" parameters
------------------------------------

Supply context parameters either in the urlpattern, or invoking get_context_data():

from `urls.py`:

.. code:: python

    urlpatterns = [
        path('daily/print/', views.ReportDailyView.as_view(), {'exclude_inactives': False}, name="daily-print"),
    ]

from a background task:

.. code:: python

    from django.core.files.base import ContentFile

    # Create a View to work with
    from reports.views import ReportDailyView
    view = ReportDailyView()
    context = view.get_context_data(
        exclude_inactives=task.exclude_inactives,
    )

    # Create empty file as result
    filename = view.build_filename(extension="weasypdf")
    task.result.save(filename, ContentFile(''))

    # Open and write result
    filepath = task.result.path

    with open(filepath, 'wb') as f:
        view.render_as_pdf_to_stream('', context, f)


Customizing the templates
-------------------------

These sample files::

    weasypdf
    ├── static
    │   └── weasypdf
    │       └── images
    │           └── header_left.png
    └── templates
        └── weasypdf
            ├── base.html
            ├── base_nomargins.html
            ├── styles.css
            ├── footer.html
            ├── header.html
            └── pages
                ├── test.css
                └── test.html


can be copied into your app's local folder `reports/templates/reports`,
and used for any required customization:

.. code:: python

    class ReportView(WeasypdfView):

        header_template_name = 'reports/header.html'
        footer_template_name = 'reports/footer.html'
        styles_template_name = 'reports/styles.css'

How to insert a page break
--------------------------

.. code:: html

    <p style="page-break-before: always" ></p>


Adding Weasyprint to your project
---------------------------------

Add `weasyprint` to your requirements::

    WeasyPrint==51

and optionally to your LOGGING setting::

    LOGGING = {
        ...
        'loggers': {
            ...
            'weasyprint': {
                'handlers': ['console'],
                'level': 'DEBUG',
                'propagate': True,
            },
        },
    }

Deployment:

1) Install Courier fonts for PDF rendering

::

    # You can verify the available fonts as follows:
    #    # fc-list
    - name: Install Courier font for PDF rendering
        become: true
        become_user: root
        copy:
            src: deployment/project/courier.ttf
            dest: /usr/share/fonts/truetype/courier/

The font file can be downloaded here:

`courier.ttf <resources/fonts/courier.ttf>`_

2) You might also need to install the following packages:

::

    #weasyprint_packages:
    - libffi-dev          # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - python-cffi         # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - python-dev          # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - python-pip          # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - python-lxml         # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - libcairo2           # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - libpango1.0-0       # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - libgdk-pixbuf2.0-0  # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - shared-mime-info    # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - libxml2-dev         # http://stackoverflow.com/questions/6504810/how-to-install-lxml-on-ubuntu#6504860
    - libxslt1-dev        # http://stackoverflow.com/questions/6504810/how-to-install-lxml-on-ubuntu#6504860

For an updated list, check here:

https://weasyprint.readthedocs.io/en/latest/install.html#linux

Customizations examples
=======================

Add a custom logo
-----------------

For example you can save a custom bitmap with django-constance:

.. code :: python

    CONSTANCE_ADDITIONAL_FIELDS = {
        'image_field': ['django.forms.ImageField', {}]
    }

    CONSTANCE_CONFIG = {
        ...
        'PDF_RECORD_LOGO': ('', 'Image for PDF logo', 'image_field'),
    }

then in your **header.html** template:

.. code:: html

    <body>
        <div class="pageHeader">
            <img class="pageLogo" title="{{ PDF_RECORD_LOGO }}" src="media://{{ PDF_RECORD_LOGO }}">
            <div class="pageTitle">{{print_date|date:'d/m/Y H:i:s'}} - {{title}}</div>
        </div>
    </body>

Embed images from media (ImageFields)
-------------------------------------

If Image is a Model to keep the images you want to embed, use a templatetag like this:

.. code:: python

    @register.filter
    def local_image_url(image_slug):
        """
        Example:
            "/backend/images/signature_mo.png"
        """

        url = ''
        try:
            image = Image.objects.get(slug=image_slug)
            if bool(image.image):
                url = image.image.url.lstrip(settings.MEDIA_URL)
        except Image.DoesNotExist as e:
            pass

        if len(url):
            url = 'media://' + url
        else:
            url = 'static://reports/images/placeholder.png'

        return url

then, in your templates:

.. code:: html

    <img class="pageLogoMiddle" src="{{'report-header-middle'|local_image_url}}">

where `'report-header-middle'` is the slug used to select the image.


Adding a plot to the PDF documents
==================================

In the frontend, you have many javascript libraries available to plot data and
draw fancy charts.

This doesn't help you in embedding a plot in a PDF documents built offline, however;
in this case, you need to build an image server side.

An helper function has been included in this app for that purpose; to use it, **matplotlib**
must be installed.

At the moment, it is more a POC then a complete solution; you can either use it
from the package, or copy the source file `weasypdf/plot.py` in your project and use
`build_plot_from_data()` as a starting point:

.. code:: python

    def build_plot_from_data(data, chart_type='line', as_base64=False, dpi=300, ylabel=''):
        """
        Build a plot from given "data";
        Returns: a bitmap of the plot

        Requires:
            matplotlib

        Keyword arguments:
        data -- see sample_line_plot_data() for an example; if None, uses sample_line_plot_data()
        chart_type -- 'line', 'bar', 'horizontalBar', 'pie', 'line', 'doughnut',
        as_base64 -- if True, returns the base64 encoding of the bitmap
        dpi -- bitmap resolution
        ylabel -- optional label for Y axis

        Data layout
        ===========

        Similar to django-jchart:

        - either (shared values for x)

            {
                "labels": ["A", "B", ...],
                "x" [x1, x2, ...],
                "columns": [
                    [ay1, ay2, ...],
                    [by1, by2, ...],
                ],
                "colors": [
                    "rgba(64, 113, 191, 0.2)",
                    "rgba(191, 64, 64, 0.0)",
                    "rgba(26, 179, 148, 0.0)"
                ]
            }

        - or

            {
                "labels": ["A", "B", ..., ],
                "columns": [
                    [
                        {"x": ax1, "y": ay1 },
                        {"x": ax2, "y": ay2 },
                        {"x": ax3, "y": ay3 },
                    ], [
                        {"x": bx1, "y": by1 },
                        {"x": bx2, "y": by2 },
                    ], ...
                ],
                "colors": ["transparent", "rgba(121, 0, 0, 0.2)", "rgba(101, 0, 200, 0.2)", ]
            }

        """

then, in the view, add the resulting bitmap to context:

.. code:: python

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        try:
            from .plot import build_plot_from_data
            plot_image = build_plot_from_data(data=None, chart_type='line', as_base64=True)
            context.update({
                'plot_image': plot_image,
            })
        except:
            pass
        return context

In the template, render it as an embedded image:

.. code:: html

    <style>
        .plot {
            border: 1px solid #ccc;
            width: 18cm;
            height: 6cm;
            margin: 1.0cm 0;
        }
    </style>

    {% if plot_image %}
        <img class="plot" src="data:image/png;base64,{{plot_image}}">
    {% endif %}

Try it
------

The management command `build_test_pdf` can be used with the "--plot_data" switch to test the resulting image:

.. code:: bash

    python manage.py build_test_pdf test.png -o -p '{"labels": ["sin", "cos"], "x": [0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5], "columns": [[0.0, 9.09, -7.57, -2.79, 9.89, -5.44, -5.37, 9.91, -2.88, -7.51], [20.0, -13.07, -2.91, 16.88, -19.15, 8.16, 8.48, -19.25, 16.68, -2.56]]}' --plot_font Tahoma

|

.. image:: screenshots/plots.png


CHANGELOG.rst
.. :changelog:

History
=======

v0.1.2
------
* PdfView renamed as WeasypdfView

v0.1.1
------
* Rename "pdf" module as "weasypdf"
* Sample project added

v0.1.0
------
* Publish on Pypi as "django-weasypdf"
* Support for 'line', 'bar', 'horizontalBar', 'pie', 'line', 'doughnut' plots

v0.0.5
------
* Support for plot added (only "line" chart type, at the moment; requires matplotlib)
* Customization examples added to readme

v0.0.4
------
* render_to_stream() renamed as render_as_pdf_to_stream()
* render_as_html_to_string() method added

v0.0.3
------
* render_to_stream() method added to produce a PDF from a background task

v0.0.2
------
* Some cleanup

v0.0.1
------
* Project start


weasypdf/utils.py
import os
from django.template.loader import get_template
from django.template.loader import render_to_string
from django.conf import settings
from django.contrib.staticfiles import finders
from django.contrib.auth import get_user_model

__author__ = 'Mario Orlandi, <mailto:morlandi@brainstorm.it>'


def trace(message):
    print(message)


################################################################################
# https://stackoverflow.com/questions/8659560/django-template-increment-the-value-of-a-variable#8671715

class Counter:
    count = 0

    def increment(self):
        self.count += 1
        return ''

    def decrement(self):
        self.count -= 1
        return ''

################################################################################
# Adapted from:
# http://ampad.de/blog/generating-pdfs-django/

def url_fetcher(url):

    import weasyprint
    original_url = url
    url = url

    try:
        if url.startswith('assets://'):
            url = url[len('assets://'):]
            with open(os.path.join(settings.ASSETS_ROOT, url), 'rb') as asset:
                contents = asset.read()
            return dict(string=contents)
        elif url.startswith('static://'):
            url = url[len('static://'):]
            with open(finders.find(url, all=False), 'rb') as asset:
                contents = asset.read()
            return dict(string=contents)

        # TODO: untested ! test me !!!
        elif url.startswith('media://'):
            url = url[len('media://'):]
            with open(os.path.join(settings.MEDIA_ROOT, url), 'rb') as asset:
                contents = asset.read()
            return dict(string=contents)

        elif url.startswith('file:///media/'):
            url = url[len('file:///media/'):]
            with open(os.path.join(settings.MEDIA_ROOT, url), 'rb') as asset:
                contents = asset.read()
            return dict(string=contents)
    except Exception as e:
        trace('Error fetching "%s"' % original_url)
        trace(str(e))
        raise

    return weasyprint.default_url_fetcher(url)


################################################################################
# Helpers

def get_pdf_styles(context, styles_template_name):
    """
    Render stylesheet from default template + (optional) custom template
    """
    styles = render_to_string('weasypdf/styles.css', context)
    if styles_template_name:
        styles += render_to_string(styles_template_name, context)
    return styles


################################################################################
# Main PDF builder

def build_pdf_document(
        base_url, debug, title, print_date, extra_context,
        styles_template_name, body_template_name, header_template_name, footer_template_name,
        output,
        format='pdf'
    ):
    """
    Create a PDF document and save it into <ouput> buffer;
    strategy:

        1) build context with initialial parameters up to <extra_context>
        2) render templates from template names
        3) build PDF content
        4) split PDF content into header, body and footer
        5) rebuild PDF document paginating previous result
    """

    def render_doc(content, base_url, styles):
        import weasyprint
        doc = weasyprint.HTML(
            string=content,
            base_url=base_url,
            url_fetcher=url_fetcher,
        ).render(
            stylesheets=[weasyprint.CSS(string=styles), ]
        )
        return doc

    def get_page_body(boxes):
        for box in boxes:
            if box.element_tag == 'body':
                return box
            return get_page_body(box.all_children())

    def get_page_fragment(template, base_url, context):
        if template:
            content = template.render(context)
            doc = render_doc(content, base_url, context['styles'])
            body = get_page_body(doc.pages[0]._page_box.all_children())
            body = body.copy_with_children(body.all_children())
        else:
            body = None
        return body

    # Build context for template rendering
    context = extra_context or {}
    context.update({
        'debug': debug,
        'format': format,
        'title': title,
        'print_date': print_date,
        'base_url': base_url,
        'MEDIA_ROOT': settings.MEDIA_ROOT,
        'STATIC_ROOT': settings.STATIC_ROOT,
    })

    # Render styles and add to context
    styles = get_pdf_styles(context, styles_template_name)
    context.update({
        'styles': styles,
    })

    # Load templates
    body_template = get_template(body_template_name) if body_template_name else None
    header_template = get_template(header_template_name) if header_template_name else None
    footer_template = get_template(footer_template_name) if footer_template_name else None

    content = body_template.render(context)
    doc = render_doc(content, base_url, context['styles'])

    # Navigate pages and Insert header and footer in main doc
    context['page_total'] = len(doc.pages)
    for i, page in enumerate(doc.pages):

        context['page_counter'] = i + 1
        page_body = get_page_body(page._page_box.all_children())

        header_body = get_page_fragment(header_template, base_url, context)
        if header_body:
            page_body.children += header_body.all_children()
        footer_body = get_page_fragment(footer_template, base_url, context)
        if footer_body:
            page_body.children += footer_body.all_children()

    doc.write_pdf(output)
    return



weasypdf/views.py
from datetime import datetime

from django.views.generic import TemplateView
from django.template.defaultfilters import slugify
from django.template.loader import get_template
from django.http import HttpResponse
from django.utils import timezone
from django.db import connection

from .utils import get_pdf_styles
from .utils import build_pdf_document
from .utils import Counter

################################################################################
# Base WeasypdfView

class WeasypdfView(TemplateView):

    body_template_name = 'weasypdf/base.html'
    styles_template_name = ''
    header_template_name = 'weasypdf/header.html'
    footer_template_name = 'weasypdf/footer.html'
    title = "Title"
    for_download = False

    body_template = None
    header_template = None
    footer_template = None

    debug = False
    format = 'pdf'
    _print_date = None

    @property
    def print_date(self):
        if self._print_date is None:
            self._print_date = timezone.now()
        return self._print_date

    def get_context_data(self, **kwargs):
        try:
            self.debug = bool(int(self.request.GET.get('debug')))
        except:
            self.debug = False
        try:
            self.format = self.request.GET.get('format')
        except:
            self.format = 'pdf'
        context = super(WeasypdfView, self).get_context_data(**kwargs)
        self.for_download = context.pop('for_download', False)
        #self.print_date = timezone.now()

        # Remove <view> from context to make sure all templates are generic
        context.pop('view', None)

        return context

    def build_filename(self, extension="pdf"):
        """
        Provide a stardard filename, build using the print datetime
        and the slugified view title;
        For example:
            '2020-04-17_19-14-55__test.pdf'
        """
        filename = timezone.localtime(self.print_date).strftime('%Y-%m-%d_%H-%M-%S__') + slugify(self.title)
        if extension:
            filename += '.' + extension
        return filename

    def render_to_response(self, context, **response_kwargs):
        if self.format == 'html':
            response = self.build_html_response(context)
        else:
            response = self.build_pdf_response(context)
        if self.for_download:
            #filename = self.print_date.strftime('%Y-%m-%d_%H-%M-%S__') + slugify(self.title) + '.pdf'
            filename = self.build_filename()
            response['Content-Disposition'] = 'attachment; filename="%s"' % filename
        return response

    def build_html_response(self, context):
        html = self.render_as_html_to_string(context)
        return HttpResponse(html)

    def render_as_html_to_string(self, context, custom_template=None):
        html = ''
        context.update({
            'debug':self.debug,
            'format':self.format,
            'title':self.title,
            'print_date': self.print_date,
            'page_total': 1,
            'page_counter': 1,
        })
        styles = get_pdf_styles(context, self.styles_template_name)
        context.update({
            'styles': styles,
        })

        if custom_template is not None:
            templates = [custom_template, ]
        else:
            templates = [self.header_template_name, self.body_template_name, self.footer_template_name, ]

        for template_name in templates:
            template = get_template(template_name) if template_name else None
            if template:
                html += template.render(context)

        # Fix unfetched assets
        html = html.replace('static://', '/static/')
        html = html.replace('media://', '/media/')
        return html

    def build_pdf_response(self, context):
        response = HttpResponse(content_type='application/pdf')
        base_url = self.request.build_absolute_uri()
        self.render_as_pdf_to_stream(base_url, context, response)
        return response

    def render_as_pdf_to_stream(self, base_url, extra_context, output):
        """
        Build the PDF document and save in into "ouput" stream.

        Automatically called when the view is invoked via HTTP (unless self.format == 'html'),
        but you can also call it explicitly from a background task:

            view = PdfTestView()
            context = view.get_context_data()
            with open(filepath, 'wb') as f:
                view.render_as_pdf_to_stream('', context, f)

        """
        build_pdf_document(
            base_url=base_url,
            debug=self.debug,
            title=self.title,
            print_date=self.print_date,
            extra_context=extra_context,
            styles_template_name=self.styles_template_name,
            body_template_name=self.body_template_name,
            header_template_name=self.header_template_name,
            footer_template_name=self.footer_template_name,
            output=output,
            format=self.format,
        )


################################################################################
# Pages

class PdfTestView(WeasypdfView):

    """
    Per il fine-tuning del css conviene invocare direttamente la view dal browser
    e selezionando "format=html",
    al fine di ottenere la pagina HTML che verrebbe utilizzata per il rendering
    del documento PDF.

    Esempio:

    http://127.0.0.1:8000/weasypdf/test/print/?format=html&debug=1

    """

    body_template_name = 'weasypdf/pages/test.html'
    styles_template_name = 'weasypdf/pages/test.css'
    # header_template_name = None
    # footer_template_name = None
    title = "Test"

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        try:
            from .plot import build_plot_from_data
            plot_image = build_plot_from_data(data=None, chart_type='line', as_base64=True)
            context.update({
                'plot_image': plot_image,
            })
        except:
            pass
        return context


sample_project/main/urls.py
"""main URL Configuration

The `urlpatterns` list routes URLs to views. For more information please see:
    https://docs.djangoproject.com/en/3.1/topics/http/urls/
Examples:
Function views
    1. Add an import:  from my_app import views
    2. Add a URL to urlpatterns:  path('', views.home, name='home')
Class-based views
    1. Add an import:  from other_app.views import Home
    2. Add a URL to urlpatterns:  path('', Home.as_view(), name='home')
Including another URLconf
    1. Import the include() function: from django.urls import include, path
    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
"""
from django.contrib import admin
from django.urls import path, include
from django.shortcuts import redirect

urlpatterns = [
    path('', lambda x: redirect('/weasypdf/test/print/')),
    path('admin/', admin.site.urls),
    path('weasypdf/', include('weasypdf.urls', namespace='weasypdf')),
    #path('reports/', include('reports.urls', namespace='reports')),
]


weasypdf/management/commands/build_test_pdf.py
# -*- coding: UTF-8 -*-
import os
import datetime
import json
import pprint
import argparse
from django.core.management.base import BaseCommand
from django.conf import settings
from weasypdf.utils import build_pdf_document
from weasypdf.views import PdfTestView
from weasypdf.plot import build_plot_from_data

        # fonts = sorted(set([f.name for f in matplotlib.font_manager.fontManager.ttflist]))
        # print(fonts)

        # rcParams['font.family'] = 'Tahoma'

class Command(BaseCommand):
    help = "Build a test PDF document (or plot)"
    epilog = """
Sample usages:


python manage.py build_test_pdf test.pdf -o


python manage.py build_test_pdf test.png -o -p '{"labels": ["sin", "cos"], "x": [0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5], "columns": [[0.0, 9.09, -7.57, -2.79, 9.89, -5.44, -5.37, 9.91, -2.88, -7.51], [20.0, -13.07, -2.91, 16.88, -19.15, 8.16, 8.48, -19.25, 16.68, -2.56]]}' --plot_font Tahoma


python manage.py build_test_pdf test.png -o --plot_type horizontalBar -p '{"labels": [null], "columns": [[1085, 839, 340, 285, 270, 211, 210, 177, 161, 141, 139, 137, 137, 134, 124, 122, 119, 116, 112, 108]], "colors": [["#f2f6ee", "#353a3a", "#fcfcfc", "#a23c3a", "#ffd100", "#0f6fa8", "#00968f", "#33b3ad", "#c80000", "#a83e3b", "#244f28", "#b8c0c2", "#586062", "#655580", "#6fd3d1", "#d0cd1d", "#cacac5", "#828b88", "#148396", "#bec8d1"]], "x": ["CW051W", "AC144N", "0011/Sugar Dust-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "AC118R", "AC105Y", "AC079N", "7116A", "7115D", "1109/Party Time-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "AC111R", "0535/Zen Retreat-EA Semibrillo SipaAlfa-Sioamundo-BaseMaster", "0546/Paternoster-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "0534/Subtle Shadow-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "7006N", "7114M", "7216N", "8792W/Monorail Silver CGl -E.A. P. y F. BioTech-Fandex Millennium-Base Blanca", "0533/Techile-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "0682/Pleasant Stream-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "0510/Sacred Spring-EA Semibrillo Sipa Alfa-Sipamundo-Base Master"]}'


python manage.py build_test_pdf test.png -o -p '{"labels": ["Total", "Custom", "Stock"], "columns": [[4, 13, 101, 78, 94, 152, 198, 337, 329, 290, 278, 295, 448, 353, 326, 512, 559, 408, 512, 440, 815, 1390, 819, 1562, 2615, 2217, 2352, 1929, 3469, 5791, 6900, 6195], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 25, 0, 0, 14, 36, 67, 8, 46, 67, 519, 585, 559, 433], [4, 13, 101, 78, 94, 152, 198, 337, 329, 290, 278, 295, 448, 353, 326, 512, 559, 408, 512, 415, 815, 1390, 805, 1526, 2548, 2209, 2306, 1862, 2950, 5206, 6341, 5762]], "colors": ["rgba(64, 113, 191, 0.2)", "rgba(191, 64, 64, 0.0)", "rgba(26, 179, 148, 0.0)"], "x": ["03/01/2018", "04/01/2018", "05/01/2018", "06/01/2018", "07/01/2018", "08/01/2018", "09/01/2018", "10/01/2018", "11/01/2018", "12/01/2018", "01/01/2019", "02/01/2019", "03/01/2019", "04/01/2019", "05/01/2019", "06/01/2019", "07/01/2019", "08/01/2019", "09/01/2019", "10/01/2019", "11/01/2019", "12/01/2019", "01/01/2020", "02/01/2020", "03/01/2020", "04/01/2020", "05/01/2020", "06/01/2020", "07/01/2020", "08/01/2020", "09/01/2020", "10/01/2020"]}'


python manage.py build_test_pdf test.png -o --plot_type pie -p '{"labels": [], "columns": [[1498, 1217, 528, 479, 363, 353]], "colors": [["rgb(255, 99, 132)", "rgb(255, 159, 64)", "rgb(255, 205, 86)", "rgb(75, 192, 192)", "rgb(54, 162, 235)", "rgb(153, 102, 255)", "rgb(201, 203, 207)"]], "x": ["255: TOO LOW WATER LEVEL", "210: DOOR OPEN", "211: COVERS NOT AVAILABLE", "2000: Paper end", "1000: Scale not responding", "895: CAN_ABSENT_DURING_DISPENSING"]}'


python manage.py build_test_pdf test.png -o --plot_type doughnut -p '{"labels": [], "columns": [[1498, 1217, 528, 479, 363, 353]], "colors": [["rgb(255, 99, 132)", "rgb(255, 159, 64)", "rgb(255, 205, 86)", "rgb(75, 192, 192)", "rgb(54, 162, 235)", "rgb(153, 102, 255)", "rgb(201, 203, 207)"]], "x": ["255: TOO LOW WATER LEVEL", "210: DOOR OPEN", "211: COVERS NOT AVAILABLE", "2000: Paper end", "1000: Scale not responding", "895: CAN_ABSENT_DURING_DISPENSING"]}'

"""

    def create_parser(self, prog_name, subcommand, **kwargs):
        if self.epilog:
            kwargs.update({
                'epilog': self.epilog,
            })
        parser = super().create_parser(prog_name, subcommand, **kwargs)
        if self.epilog:
            parser.formatter_class =  argparse.RawTextHelpFormatter
        return parser

    def add_arguments(self, parser):
        parser.add_argument("filepath", help='ouput file path (use .pdf or .png extension as required')
        parser.add_argument('--open', '-o', action='store_true', help="open resulting file")
        parser.add_argument("--plot_data", "-p", help='JSON plot data to build a bitmap (instead of a pdf)')
        parser.add_argument('--plot_type', "-t", choices=['line', 'bar', 'horizontalBar', 'pie', 'doughnut', ], default="line")
        parser.add_argument('--plot_font', )
        parser.add_argument('--list_fonts', "-l", action='store_true', help="list matplotlib fonts and exit")

    def list_fonts(self):
        import matplotlib
        fonts = list(set([f.name for f in matplotlib.font_manager.fontManager.ttflist]))
        sorted_fonts = sorted(fonts, key=str.casefold)
        pprint.pprint(sorted_fonts)

    def select_font(self, font):
        import matplotlib
        matplotlib.rcParams['font.family'] = font

    def handle(self, *args, **kwargs):

        if kwargs['list_fonts']:
            # list fonts and exit
            self.list_fonts()
            return

        if kwargs['plot_font']:
            self.select_font(kwargs['plot_font'])

        # Use either user supplied filepath, or the a default name provided by the view
        expected_file_extension = '.pdf'
        filepath = kwargs['filepath']
        filename, file_extension = os.path.splitext(filepath)

        # Retrieve and deserialized user-supplied plot data
        plot_data = kwargs['plot_data']
        if plot_data is not None:
            plot_data = json.loads(plot_data)
            expected_file_extension = '.png'

        # File cleanup
        assert file_extension.lower() == expected_file_extension, "Wrong extension: %s" % file_extension
        if os.path.isfile(filepath):
            os.remove(filepath)

        with open(filepath, 'wb') as f:
            if plot_data is not None:
                image = build_plot_from_data(
                    plot_data,
                    chart_type=kwargs['plot_type'],
                    as_base64=False,
                    dpi=300,
                    ylabel='test plot',
                )
                f.write(image)
            else:
                # Create a View to work with
                view = PdfTestView()
                context = view.get_context_data()
                view.render_as_pdf_to_stream('', context, f)

        print('Result in: "%s"' % os.path.abspath(filepath))
        if kwargs['open']:
            os.system('open %s' % filepath)
